crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,498 @@
+    let ast::ExprIf {
+        test,
+        body,
+        orelse,
+        range: _,
+    } = if_expr;
+
+    let ast::ExprSubscript { value, slice, .. } = body.as_subscript_expr()?;
+    let else_or_target_name = &orelse.as_name_expr()?.id;
+    // Variable names correspond to:
+    // ```python
+    // value[slice] if test else else_or_target_name
+    // ```
+    affix_removal_data_core_logic(value, test, else_or_target_name, slice)
+}
+
+/// Given a statement of the form:
+///
+/// ```python
+///  if text.func(affix):
+///     text = text[slice]
+/// ```
+///
+/// where `func` is either `startswith` or `endswith`,
+/// this function collects `text`,`func`, `affix`, and the non-null
+/// bound of the slice. Otherwise, returns `None`.
+fn affix_removal_data_stmt(if_stmt: &ast::StmtIf) -> Option<RemoveAffixData> {
+    let ast::StmtIf {
+        test,
+        body,
+        elif_else_clauses,
+        range: _,
+    } = if_stmt;
+
+    // Cannot safely transform, e.g.,
+    // ```python
+    // if text.startswith(prefix):
+    //     text = text[len(prefix):]
+    // else:
+    //     text = "something completely different"
+    // ```
+    if !elif_else_clauses.is_empty() {
+        return None;
+    };
+
+    // Cannot safely transform, e.g.,
+    // ```python
+    // if text.startswith(prefix):
+    //     text = f"{prefix} something completely different"
+    //     text = text[len(prefix):]
+    // ```
+    if body.len() != 1 {
+        return None;
+    };
+
+    // Variable names correspond to:
+    // ```python
+    // if test:
+    //     else_or_target_name = value[slice]
+    // ```
+    let ast::StmtAssign {
+        value,
+        targets,
+        range: _,
+    } = body[0].as_assign_stmt()?;
+    if targets.len() != 1 {
+        return None;
+    };
+    let else_or_target_name = &targets[0].as_name_expr()?.id;
+    let ast::ExprSubscript { value, slice, .. } = value.as_subscript_expr()?;
+
+    affix_removal_data_core_logic(value, test, else_or_target_name, slice)
+}
+
+/// Suppose given a statement of the form:
+/// ```python
+/// if test:
+///     else_or_target_name = value[slice]
+/// ```
+/// or an expression of the form:
+/// ```python
+/// value[slice] if test else else_or_target_name
+/// ```
+/// This function verifies that
+///     - `value` and `else_or_target_name`
+/// are equal to a common name `text`
+///     - `test` is of the form `text.startswith(prefix)`
+/// or `text.endswith(suffix)`
+///     - `slice` has no upper bound in the case of a prefix,
+/// and no lower bound in the case of a suffix
+///
+/// If these conditions are satisfied, the function
+/// returns the corresponding `RemoveAffixData` object;
+/// otherwise it returns `None`.
+fn affix_removal_data_core_logic<'a>(
+    value: &'a ast::Expr,
+    test: &'a ast::Expr,
+    else_or_target_name: &'a ast::name::Name,
+    slice: &'a ast::Expr,
+) -> Option<RemoveAffixData<'a>> {
+    let body_name = &value.as_name_expr()?.id;
```

Comment:
I wonder if we _need_ to exclude other valid assignment targets, like attribute expressions. We could possibly use our [`ComparableExpr`](https://github.com/astral-sh/ruff/blob/a7c936878db8d9906c6ca91c8a1301aee80ee78a/crates/ruff_python_ast/src/comparable.rs#L1-L16) struct to facilitate comparisons of other kinds of assignment targets. It shouldn't really matter what kind of assignment target it is, as long as it's the same in all three places -- i.e. this should be fine for us to flag and fix:

```py
import foo.bar

SUFFIX = "suffix"

x = foo.bar.baz[:len(SUFFIX)] if foo.bar.baz.endswith(SUFFIX) else foo.bar.baz
```

Reply:
Rule expanded! (there's a typo in your example though: the slice needs a `-` preceding `len`). Test fixture updated as well.